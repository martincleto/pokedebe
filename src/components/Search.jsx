
import React, { Component } from 'react'
import Autosuggest from 'react-autosuggest'
import {hashHistory} from 'react-router'

import styles from 'Stylesheets/search.scss'  // eslint-disable-line no-unused-vars

class Search extends Component {

  constructor(props) {
    super(props)

    this.state = {
      value: '',
      suggestions: []
    }
  }

  getSuggestions(value) {
    const inputValue = value.trim().toLowerCase()
    const inputLength = inputValue.length
    const data = this.props.data.results

    return inputLength === 0 ? [] : data.filter(item =>
      item.name.toLowerCase().slice(0, inputLength) === inputValue
    )
  }

  getSuggestionValue(suggestion) {
    return suggestion.name
  }

  renderSuggestion(suggestion) {
    return (
      <div>{suggestion.name}</div>
    )
  }

  onChange(event, { newValue }) {
    this.setState({
      value: newValue
    })
  }

  onSearch(event) {
    event.preventDefault()

    hashHistory.push(`/detail/${this.state.value}`)
  }

  onSuggestionsFetchRequested({ value }) {
    let suggestions = this.getSuggestions(value)

    this.setState({
      suggestions: suggestions
    })
  }

  onSuggestionsClearRequested() {
    this.setState({
      suggestions: []
    })
  }

  render() {
    const { value, suggestions } = this.state
    const inputProps = {
      placeholder: 'Enter a Pokemon name e.g. Bulbasaur',
      value,
      onChange: this.onChange.bind(this)
    }

    return (
      <form className="search-form">
        <Autosuggest
          suggestions={suggestions}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested.bind(this)}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested.bind(this)}
          getSuggestionValue={this.getSuggestionValue}
          renderSuggestion={this.renderSuggestion}
          inputProps={inputProps}
        />
      <button id="btn-search" className="search-form__button" onClick={this.onSearch.bind(this)}>Search!</button>
      </form>
    )
  }
}

Search.propTypes = {
  data: React.PropTypes.object
}

export default Search
