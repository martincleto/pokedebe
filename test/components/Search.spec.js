import React from 'react'
import renderer from 'react-test-renderer'
import Search from 'Components/Search'

describe('Search', function() {
  test('should render without error', () => {
    const component = renderer.create(
      <Search/>
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
})
