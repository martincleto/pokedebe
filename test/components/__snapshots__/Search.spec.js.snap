// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render without error 1`] = `
<p>
  Search
</p>
`;
